src/log.h

```diff
@@ -221,16 +279,62 @@ class StreamProxy {
 public:
 	StreamProxy(std::ostream *os) : m_os(os) { }
 
+	inline static void maybe_fix_stream_state(std::ostream &os)
+	{
+		std::ios::iostate state = os.rdstate();
+		if (state != std::ios::goodbit) {
+			// clear error state so the stream works again
+			os.clear();
+			if (state & std::ios::eofbit)
+				os << "(ostream:eofbit)";
+			if (state & std::ios::badbit)
+				os << "(ostream:badbit)";
+			if (state & std::ios::failbit)
+				os << "(ostream:failbit)";
+		}
+	}
+
 	template<typename T>
-	StreamProxy& operator<<(T&& arg) {
+	StreamProxy& operator<<(T&& arg)
+	{
 		if (m_os) {
+			maybe_fix_stream_state(*m_os);
 			*m_os << std::forward<T>(arg);
 		}
 		return *this;
 	}
 
-	StreamProxy& operator<<(std::ostream& (*manip)(std::ostream&)) {
+	template<typename T>
+	StreamProxy& emit_with_null_check(T&& arg)
+	{
+		if (arg == NULL)
+			return *this << "(null)";
+		else
+			return this->operator<< <T>(std::forward<T>(arg));
```

Comment:
couldn't this resolve back to the below functions?

Reply:
The first works because it is actually invoking `this->operator<< <char[7]>`, which bypasses the null checking. I will make it explicit with a comment and cast.

The second forces the use of the template `operator<< <T>` rather than overloaded `operator<<`. The methods produced by `ADD_NULL_CHECK` are regular method overloads rather than template specializations. These are preferred when the type matches exactly.

Reply:
I will add more cases to the unit test to make sure every overload is tested.